R/series__series.R

```diff
@@ -61,34 +93,104 @@
 #' s_copy = s
 #' .pr$Series$append_mut(s, pl$Series(5:1))
 #' identical(s_copy$to_r(), s$to_r()) # s_copy was modified when s was modified
+#'
+#' # call functions via sub-namespaces
+#' pl$Series(c(1:3))$expr$add(1)
+#'
+#' pl$Series(list(3:1, 1:2, NULL))$list$first()
+#'
+#' pl$Series(c(1, NA, 2))$str$concat("-")
 NULL
 
 
-Series_flags = method_as_property(function() {
-  out = list(
-    "SORTED_ASC" = .pr$Series$is_sorted_flag(self),
-    "SORTED_DESC" = .pr$Series$is_sorted_reverse_flag(self)
-  )
+# Active bindings
+
+Series_dtype = method_as_active_binding(\() .pr$Series$dtype(self))
+
+
+Series_flags = method_as_active_binding(
+  \() {
+    out = list(
+      "SORTED_ASC" = .pr$Series$is_sorted_flag(self),
+      "SORTED_DESC" = .pr$Series$is_sorted_reverse_flag(self)
+    )
+
+    # the width value given here doesn't matter, but pl$Array() must have one
+    if (pl$same_outer_dt(self$dtype, pl$List()) ||
+      pl$same_outer_dt(self$dtype, pl$Array(width = 1))) {
+      out[["FAST_EXPLODE"]] = .pr$Series$fast_explode_flag(self)
+    }
 
-  # the width value given here doesn't matter, but pl$Array() must have one
-  if (pl$same_outer_dt(self$dtype, pl$List()) ||
-    pl$same_outer_dt(self$dtype, pl$Array(width = 1))) {
-    out[["FAST_EXPLODE"]] = .pr$Series$fast_explode_flag(self)
+    out
   }
-  out
-})
+)
 
-#' String related methods
-#'
-#' Create an object namespace of all string related methods. See the individual
-#' method pages for full details.
-#' @return Series
-#' @noRd
-Series_str = method_as_property(function() {
-  series_str_make_sub_ns(self)
+
+Series_name = method_as_active_binding(\() .pr$Series$name(self))
+
+
+Series_shape = method_as_active_binding(\() .pr$Series$shape(self))
+
+
+# Sub-namespaces
+
+Series_expr = method_as_active_binding(function() {
+  df = pl$select(self) # make a DataFrame from Series
+  self = pl$col(self$name) # override self to column named by series
+
+  # loop over each expression function
+  lapply(
+    RPolarsExpr,
+    \(f) { # f is orignial Expr method
+
+      # point back to env with above defined 'df' and 'self'
+      environment(f) = parent.frame(2L)
+
+      # make a modified Expr function
+      new_f = \() {
+        # get the future args the new function will be called with
+        # not using ... as this will erase tooltips and defaults
+        # instead using sys.call/do.call
+        scall = as.list(sys.call()[-1])
+
+        x = do.call(f, scall)
+        pcase(
+          inherits(x, "RPolarsExpr"), df$select(x)$to_series(0),
+          or_else = x
+        )
+      }
+
+      # set new_f to have the same formals arguments
+      formals(new_f) = formals(f)
+      class(new_f) = c("SeriesExpr", "function") #
+      new_f
+    }
+  )
 })
 
 
+Series_list = method_as_active_binding(
+  \() {
+    df = pl$select(self)
+    arr = expr_list_make_sub_ns(pl$col(self$name))
+    lapply(arr, \(f) {
+      \(...) df$select(f(...))
```

Comment:
I think adding `$to_series()` here would be enough to address my previous comment